content/articles/django-model-polymorphism/index.md

```diff
@@ -0,0 +1,226 @@
+description: This tutorial will go over the concept of Django model design from the default feature provided by Django to more advanced modelling features available in Django
+author: flavian-adhiambo
+date: 2022-02-05T00:00:00-05:41
+topics: [API]
+excerpt_separator: <!--more-->
+images:
+
+  - url: /engineering-education/django-model-polymorphism/hero.png
+    alt: Getting started with Django polymorphic models
+---
+
+### Django polymorphic database modelling
+Polymorphism is a technique for modelling data in a database so that applications can use it to store different types of data in the same table. For example, in an online bookstore, a user can order an ebook version of a novel while another user orders a print version of the same book. In the database, it's the same book but with different properties. The ebook version will include a download link which the print version won't have. The print version can have a weight property that we can use to calculate the ebook version's delivery charges.
+
+### Table of contents
+- [Django polymorphic database modelling](#django-polymorphic-database-modelling)
+- [Table of contents](#table-of-contents)
+- [Prerequisites](#prerequisites)
+- [Polymorphic Modelling](#polymorphic-modelling)
+  - [Default modeling](#default-modeling)
+  - [Sparse Modelling](#sparse-modelling)
+  - [Semi-structured modelling](#semi-structured-modelling)
+  - [Abstract base modelling](#abstract-base-modelling)
+- [Conclusion](#conclusion)
+  
+### Prerequisites
+1. [Python](https://www.python.org/) and [Virtualenv](https://virtualenv.pypa.io/en/latest/) installed in your computer.
+2. Knowledge in [Django](https://docs.djangoproject.com/en/4.0/intro/overview/) and [Python](https://www.python.org/).
+   
+### Polymorphic Modelling
+Django ORM provides several ways to model polymorphic data. As we will study below, we can either use the standard Django features or the advanced Django ORM features. As we will see below, there are several ways to model polymorphic data. We will begin with the simplest method to the complex way of implementing polymorphic models.
+
+This tutorial will model an online book store data dealing with print books and ebooks sales. We will begin by creating simple models and adding more fields to our models as the online store expands and handles more products.
+
+We will start by creating a new Django project that we will use throughout this guide. 
+
+1. Create a new working directory name `models` by executing the command below on the command line.
+   ```bash
+    mkdir modelling
+   ```
+2. Execute the command below to change the working directory to the one we have created above and create a new virtual environment. Django projects use virtual environments to manage the project's dependencies in isolations to avoid dependency conflict between different projects.
+   ```bash
+   virtualenv venv
+   source venv/bin/activate
+   pip install django
+   ```
```

Comment:
Delete extra spaces.

Reply:
@ahmadmardeni1 the extra space is meant to align the code section with the current step.